### PR TITLE
Pushing site_cmp3 statistics to prometheus pushgateway

### PR DESCRIPTION
**Pull request by eachristgr**
Monitoring consistency check jobs with newer and more suitable technologies is essential. Thus, the pull request implemented allow to a part of the consistency check (site_cmp3) pushing metrics to a prometheus pushgateaway. 

For the implementation, the [MetricManager](https://github.com/rucio/rucio/blob/0bbc069faf17f06eff1891aa90e634adea67b216/lib/rucio/core/monitor.py#L327) from [Rucio](https://github.com/rucio/rucio) was used, in order to align the monitoring of consistency check with the monitoring of other CMS Rucio relevant tools.

This implementation can easily integrate other parts of consistency check in the future, allowing even more detailed monitoring.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -157,4 +157,6 @@ cython_debug/
 #  be found at https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
-#.idea/
\ No newline at end of file
+#.idea/
+
+prometheus_dev
```
Comment (dynamic-entropy) [inline .gitignore]: Its a bit non standard directory. Unless you have development instructions that use this, its better to add a gitignore in the directory and ignore everything inside it.

File: docker/Dockerfile [hunk 2]
```diff
@@ -53,14 +53,15 @@ RUN dnf install -y python3 python3-pip git \
 RUN pip3 install --upgrade pip
 RUN pip3 install --upgrade setuptools
 RUN pip3 --no-cache-dir install SQLAlchemy pyyaml pythreader cx_Oracle j2cli
+RUN python3 -m pip install --no-cache-dir --pre rucio[oracle,mysql,postgresql]
```
Comment (dynamic-entropy) [inline docker/Dockerfile]: For consistency, can this not be installed via pip and clubbed with the below install call?

File: site_cmp3/site_cmp3.sh [hunk 3]
```diff
@@ -224,7 +225,19 @@ ed_action_errors=${out}/${RSE}_${now}_ED_action.errors
...
+}
 _EOF_
+
+#
+# 7. Push2Prometheus
+#
+echo
+echo Pushing stats...
+echo
+$python /consistency/push2prometheus.py $stats site_cmp3
```
Comment (dynamic-entropy) [inline site_cmp3/site_cmp3.sh]: Later, do you think errors from this should be piped to a directory as well or do you prefer to keep them in stdout?

**Review by dynamic-entropy — COMMENTED** [review]
I am yet to go through `docker/push2prometheus.py`. These are non-functional comments.

Otherwise, it looks good and ready for testing.

File: docker/Dockerfile (hunk 2 above)
Comment (eachristgr) [reply docker/Dockerfile]: pip3 can be used directly, although `--pre` flag is also used in similar situations. As there is no other package being installed using this flag, rucio can not be cluubed with other packages.

File: site_cmp3/site_cmp3.sh (hunk 3 above)
Comment (eachristgr) [reply site_cmp3/site_cmp3.sh]: Subprocesses, triggered by this script, are called by redirecting their stderr to certain files. This file are placed under `/var/cache/consistency-dump` directory of the running container and can be used for debugging. The retention policy applied to these files is specified in [cleanup.sh](https://github.com/eachristgr/cms_consistency/blob/prometheus/docker/cleanup.sh).

**Comment by eachristgr** [comment]
@dynamic-entropy changes are applied.

File: docker/Dockerfile (hunk 2 above)
Comment (dynamic-entropy) [reply docker/Dockerfile]: Why do you want to use the `pre-release` package?

File: site_cmp3/site_cmp3.sh (hunk 3 above)
Comment (dynamic-entropy) [reply site_cmp3/site_cmp3.sh]: Does this call `site_cmp3`?

**Review by dynamic-entropy — COMMENTED** [review]
What is scan_type? why is a command being passed to the prometheus script?

Isn't it supposed to simply scrape the result jsons and push to the push gateway?

**Comment by eachristgr** [comment]
> What is scan_type? why is a command being passed to the prometheus script?
> 
> Isn't it supposed to simply scrape the result jsons and push to the push gateway?

It is not a function call. site_cmp3 is handled like a string inside the python script. However, I changed it so there is no more confusion.

In details, site_cmp3 is the name of the script that includes the command. site_cmp3 triggers the required subprocesses for consistency check on sites merged volume. Different scripts that trigger different subprocesses are used for the unmerged volume of the sites and RAL site. I believe that allowing prometheus script to take the script name as an argument (scan_type), will help to easily integrate in the future the rest of consistency check in the monitoring

**Pull request closed without merging**